DDF-2334 Null values displayed in metacard organization attributes

vinamartin opened the pull request:
https://github.com/codice/ddf/pull/1109

Provides null checks within metacard attributes. For instance, the phone number area code would return "(null) 123-4353" instead of "123-4353". The phone number and address metacard attributes now have more null checks.

@coyotesqrl
@lessarderic

coyotesqrl commented:
Please copy the description from the #1109 PR into your description.

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -490,24 +493,39 @@ private static void setMetacardEmailAttribute(List<EmailAddressType> emailAddres
-                        digits.getAreaCode(),
-                        digits.getNumber()));
+    private static StringBuilder buildNonNullString(StringBuilder stringBuilder,
+            String attributeValue, String format, String delimiterBefore) {
```
Feels weird to have the potentially empty value be penultimate.

vinamartin replied:
Could you expand on this comment some more?

coyotesqrl replied:
The overloaded method invokes this by passing in an empty string to `format` (which I've commented should probably be `null` instead, but that's a separate issue). It's more common when you've got an overloaded private method that the optional/null values be passed in at the end.

vinamartin replied:
Oh gotcha, yeah, now that I look at it this is pretty strange. I'll fix this up. 
As for your other comment, could you expand on that some more as well?

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -490,24 +493,39 @@ private static void setMetacardEmailAttribute(List<EmailAddressType> emailAddres
+    private static StringBuilder buildNonNullString(StringBuilder stringBuilder,
+            String attributeValue, String delimiterBefore) {
+
+        return (buildNonNullString(stringBuilder, attributeValue, "", delimiterBefore));
```
Seems weird to me in a private method not to use `null` as your "I'm not here" flag.

coyotesqrl replied:
Expanding: using an empty string for an internal method as a flag that "I should be ignored" seems strange. A null would be the more common indicator.

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -437,36 +438,38 @@ private static boolean isAttributeMultiValued(String attribute, MetacardImpl met
-        }
-
-        return phoneNumber;
+    private static String getPhoneNumber(TelephoneNumberType digits) {
```
So we only support US phone numbers? No country codes?

vinamartin replied:
Weird, we are supporting country codes but as you have pointed out we are not building it here. 
I'm fixing that up now. 

What are your thoughts on how that should be displayed?
I was thinking the following:

+1 (555) 555-5555 extension 555

coyotesqrl replied:
That works for me. 

I'm assuming the area code/prefix/number/extension parts are pretty universal.

lessarderic replied:
`extension` makes the phone number fairly long and take a lot of space. Should we consider using `ext` instead?

vinamartin replied:
Works for me. That's how we've been displaying it in the report anyway.

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -490,24 +493,39 @@ private static void setMetacardEmailAttribute(List<EmailAddressType> emailAddres
 
     private static void setMetacardPhoneNumberAttribute(List<TelephoneNumberType> phoneNumbers,
             String metacardAttribute, MetacardImpl metacard) {
+
```
This should work and be a smidge clearer.
```java
if (phoneNumbers != null) {
    List<String> metacardPhoneNumbers = phoneNumbers.stream()
        .map(this::getPhoneNumber)
        .collect(Collectors.toList());
    metacard.setAttribute(metacardAttribute, (Serializable) metacardPhoneNumbers);
}

coyotesqrl commented:
Are there any negative tests? I don't see any where we have empty phone number/address.

coyotesqrl commented:
Small nits aside (and as these changes are being cherry picked from master I can certainly ignore them on this branch), 👍

vinamartin commented:
@coyotesqrl I'm going to go ahead and address these comments on master

vinamartin commented:
@coyotesqrl Could I get a re-review on this? I have addressed most of your comments for now. I'll cherry-pick this back and make a PR into master once I get a 👍

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -437,36 +438,43 @@ private static boolean isAttributeMultiValued(String attribute, MetacardImpl met
+        phoneNumberBuilder = buildNonNullString(phoneNumberBuilder,
+                digits.getExtension(),
+                " ",
+                "extension %s");
```
I just want to throw in a comment here - I don't expect it to be addressed - that this is the sort of thing that kills us with I18N/L10N.

vinamartin replied:
I don't disagree. :)

lessarderic replied:
We may want to create a constant for the string we use for extension.

coyotesqrl commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -437,36 +438,43 @@ private static boolean isAttributeMultiValued(String attribute, MetacardImpl met
-
-            if (StringUtils.isNotBlank(metacardAddress)) {
-                metacard.setAttribute(metacardAttribute, metacardAddress);
+            addressString = buildNonNullString(addressString, address.getStreet(), " ");
```
Do you want the space character prefixing the first line of the address here?

coyotesqrl replied:
Ah...never mind. When the StringBuilder has length 0 it's not used. Never mind.

coyotesqrl commented:
👍

lessarderic commented on catalog/spatial/registry/registry-ebrim-transformer/src/main/java/org/codice/ddf/registry/converter/RegistryPackageConverter.java:
```diff
@@ -437,36 +438,43 @@ private static boolean isAttributeMultiValued(String attribute, MetacardImpl met
+                digits.getAreaCode(),
+                " ",
+                "(%s)");
+        phoneNumberBuilder = buildNonNullString(phoneNumberBuilder, digits.getNumber(), " ");
```
Does it even make sense to create a phone number string if the number is missing? Should we consider returning a blank string in that case?

vinamartin replied:
The UI has validation to prevent someone from putting in a phone number like this.
That should prevent that from happening.

coyotesqrl replied:
Never rely on UI validations.

vinamartin replied:
Again, I don't disagree. :)

lessarderic commented on catalog/spatial/registry/registry-ebrim-transformer/src/test/java/org/codice/ddf/registry/transformer/RegistryTransformerTest.java:
```diff
@@ -118,26 +118,52 @@ public void testBasicInfo() throws Exception {
+        assertThat(RegistryUtility.getStringAttribute(metacard,
+                RegistryObjectMetacardType.ORGANIZATION_NAME,
+                null), is("Codice"));
+        assertThat(RegistryUtility.getListOfStringAttribute(metacard,
```
We're not testing any of the scenarios where a piece of the address is `null` or missing. We're also not testing many of the other scenarios such as what if the extension isn't there.

Maybe we should extract `StreetAddressStringBuilder` and `PhoneNumberStringBuilder` classes and unit test those independently. That way you wouldn't have to do that low-level testing here.

lessarderic commented:
Just realized those are coming from a 2.10.0 PR that's already been merged so we may need to port those to the 2.10 branch as well :-(

👍  when the comments have been addressed.

vinamartin commented:
Made a ticket for the testing tech debt: https://codice.atlassian.net/browse/DDF-2419 Will address this soon. 

I'll be cherry-picking this back over to master, as well.

adimka commented:
Hero'd succesfully 👍
